# Solving Nonlinear Equations

To solve a nonlinear equation, we used the package ``NLsolve``.

In [1]:
using NLsolve

Our problem is to solve for $r$ in

$$
   1 - e^{-25 r} = 12.5 \left( 1 - e^{-r} \right).
$$

The above equation is equivalent to

$$
   1 - e^{-25 r} - 12.5 \left( 1 - e^{-r} \right) = 0.
$$

Writing the above equation as $f(x) = 0$, the function below defines $f$ as the input to ``nlsolve`` of the package ``NLsolve``.

In [2]:
"""
    function f!(F, x)

defines F[1] as the input to nlsolve.
"""
function f!(F, x)
   r = x[1]
   F[1] = 1 - exp(-25*r) - 12.5*(1 - exp(-r)) 
end

f!

Let us pass $r = 0.05$ as the initial value for the solver.

In [3]:
sol = nlsolve(f!, [0.05])

Results of Nonlinear Solver Algorithm
 * Algorithm: Trust-region with dogleg and autoscaling
 * Starting Point: [0.05]
 * Zero: [0.0673745373743581]
 * Inf-norm of residuals: 0.000000
 * Iterations: 4
 * Convergence: true
   * |x - x'| < 0.0e+00: false
   * |f(x)| < 1.0e-08: true
 * Function Calls (f): 5
 * Jacobian Calls (df/dx): 5

We see that the solution is 0.0673745373743581.

Let us verify.

In [4]:
f(r) = 1 - exp(-25*r) - 12.5*(1 - exp(-r))

f (generic function with 1 method)

Let us see what the output type of ``nlsolve`` really is.

In [5]:
dump(sol)

NLsolve.SolverResults{Float64, Float64, Vector{Float64}, Vector{Float64}}
  method: String "Trust-region with dogleg and autoscaling"
  initial_x: Array{Float64}((1,)) [0.05]
  zero: Array{Float64}((1,)) [0.0673745373743581]
  residual_norm: Float64 9.36140054363932e-13
  iterations: Int64 4
  x_converged: Bool false
  xtol: Float64 0.0
  f_converged: Bool true
  ftol: Float64 1.0e-8
  trace: NLsolve.SolverTrace
    states: Array{NLsolve.SolverState}((0,))
  f_calls: Int64 5
  g_calls: Int64 5


In [6]:
sol.zero

1-element Vector{Float64}:
 0.0673745373743581

In [7]:
f(sol.zero[1])

-9.36140054363932e-13

The residual is very small, so we accept the solution.